### PR TITLE
Add documentation: internals, contributing, development

### PR DESCRIPTION
**Pull request by pierDipi**
Signed-off-by: Pierangelo Di Pilato <pierangelodipilato@gmail.com>

## Proposed Changes

- Add `INTERNALS.md` for documenting the overall internal architecture.
- Add `DEVELOPMENT.md` explaining the basic development cycle.
- Add `data-plane/README.md` explaining data-plane architecture and directory structure.
- Add `CONTRIBUTING.md` same as `eventing` with some custom links.

### CODE REVIEW

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+_Adding the `upstream` remote sets you up nicely for regularly
+[syncing your fork](https://help.github.com/articles/syncing-a-fork/)._
+
+Once you reach this point you are ready to do a full build and deploy as
+follows.
+
+# Deploy core configurations and Kafka
+
+```bash
+# re-execute the script in case some errors appear.
```
Comment (slinkydeveloper) [inline DEVELOPMENT.md:78]: In a followup, Could you have a retry inside the script more than asking to the user to do so?

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    The script contains instructions on how to use it, and what it does.
+    
+If you are using [KinD](https://kind.sigs.k8s.io/) we recommend executing:
+
+```bash
+export WITH_KIND=true
+export SKIP_PUSH=true
+```
+
+This loads images in KinD and skips the push to the remote registry pointed by `KO_DOCKER_REPO`, allowing speeding up
```
Comment (slinkydeveloper) [inline DEVELOPMENT.md:97]: cool! In a followup could we have it for minikube too?

File: INTERNALS.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# Architecture
+
+The document [Alternative Broker Implementation based on Apache Kafka](https://docs.google.com/document/d/10-qylWrj7Tj81EqoIiZ2AANAZNmkKe7XKmFobgZULKY/edit)
+explains the architecture, and some reasons for implementing a native Kafka Broker instead of using a channel-based Broker. 
+(You need to join `knative-dev` group to view it).
```
Comment (slinkydeveloper) [inline INTERNALS.md:5]: More than linking the doc, could you report here the major things explained in the doc?

File: INTERNALS.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+The **data-plane** is implemented in Java for leveraging the always up-to-date, feature rich and tuned Kafka client.
+The **control-plane** is implemented in Go following the 
+[Knative style of writing controllers](https://github.com/knative-sandbox/sample-controller).
```
Comment (slinkydeveloper) [inline INTERNALS.md:9]: s/Knative style of writing controllers/Knative Kubernetes controllers

File: data-plane/README.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# Data-plane
+
+The data-plane uses [Vertx](https://vertx.io/) and is composed of two components:
+- [**Receiver**](#Receiver), it's responsible for accepting incoming events and send them to the appropriate Kafka topic.
```
Comment (slinkydeveloper) [inline data-plane/README.md:4]: the link to anchors should be lowercase IIRC

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+1. [Create and checkout a repo fork](#checkout-your-fork)
+
+Before submitting a PR, see also [contribution guidelines](./CONTRIBUTING.md).
+
+### Requirements
+
+You must have installed:
+
+- [`ko`](https://github.com/google/ko)
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply DEVELOPMENT.md:18]: Go yes, Java and Maven are optional (a developer can use the script `hack/dev_data_plane_setup.sh`, which is docker-based, for running data-plane unit tests). Of course, if you install them, the whole process is easier. 

Let me clarify this point and add Go.

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply DEVELOPMENT.md:18]: I think u should add Java and Maven too, this is a development document explaining how to develop, so to me it makes sense expliciting the tools you need to develop (you need a jdk to index the ide and run locally the tests, even if for your scripts you don't need them)

File: DEVELOPMENT.md (hunk 2 above)
Comment (pierDipi) [reply DEVELOPMENT.md:97]: Of course, I would love to have minikube as a first-class development environment as well.

File: data-plane/README.md (hunk 5 above)
Comment (pierDipi) [reply data-plane/README.md:4]: I think you are right, thanks

File: DEVELOPMENT.md (hunk 1 above)
Comment (pierDipi) [reply DEVELOPMENT.md:78]: Good point! I opened an issue: https://github.com/knative-sandbox/eventing-kafka-broker/issues/12

File: INTERNALS.md (hunk 3 above)
Comment (pierDipi) [reply INTERNALS.md:5]: Can we do this when we consolidate the implementation after we implement the `control-plane`?

Comment (slinkydeveloper) [reply INTERNALS.md:5]: Sure, open an issue for that :wink:

Comment (pierDipi) [reply INTERNALS.md:5]: created: https://github.com/knative-sandbox/eventing-kafka-broker/issues/13

**Pull request merged**